openlibrary/core/bookshelves.py

```diff
@@ -202,6 +203,101 @@ def count_total_books_logged_by_user_per_shelf(
         result = oldb.query(query, vars=data)
         return {i['bookshelf_id']: i['count'] for i in result} if result else {}
 
+    # Iterates through a list of solr docs, and for all items with a 'logged edition'
+    # it will remove an item with the matching edition key from the list, and add it to
+    # doc["editions"]["docs"]
+    def link_editions_to_works(solr_docs):
+        linked_docs: list[web.storage] = []
+        editions_to_work_doc = {}
+        # adds works to linked_docs, recording their edition key and index in docs_dict if present.
+        for doc in solr_docs:
+            if doc["key"].startswith("/works"):
+                linked_docs.append(doc)
+                if doc.get("logged_edition"):
+                    editions_to_work_doc.update({doc["logged_edition"]: doc})
+        # Attaches editions to the works, in second loop-- in case of misperformed order.
+        for edition in solr_docs:
+            if edition["key"].startswith("/books/"):
+                if work_doc := editions_to_work_doc.get(edition["key"]):
+                    work_doc.editions = [edition]
+                else:
+                    # raise error no matching work found? Or figure out if the solr queries are performedorderless.
+                    logger.warning(
```

Comment:
This should never happen, but logger.error will appear in sentry 

```suggestion
                    logger.error(
```